### PR TITLE
NITF: Update for SNIP TREs

### PR DESCRIPTION
**Pull request by crossbearer**
A new version of NITF will be available soon (definition file attached from https://nsgreg.nga.mil/doc/view?i=4829). This file has TREs with field value types of IEEE754 Float and UINT as opposed to the traditional string parsing.
[NGA.STND.0072_1.0_SNIP_Version_1.0_07-June-2019(1).pdf](https://github.com/OSGeo/gdal/files/4370390/NGA.STND.0072_1.0_SNIP_Version_1.0_07-June-2019.1.pdf)
I ended up redoing this in order to have nitffile.c act correctly in the comparison tool.

### CODE REVIEW

File: gdal/frmts/nitf/nitffile.c [hunk 1]
```diff
@@ -2216,7 +2217,77 @@ static char** NITFGenericMetadataReadTREInternal(char **papszMD,
                 (*pnMDSize) ++;
                 papszTmp[0] = NULL;
                 CPLFree(papszTmp);
+                const char* pszVal = strchr(papszMD[(*pnMDSize) - 1], '=') + 1;
+                
+				if (strcmp(pszType, "IEEE754") == 0)
```
Comment (rouault) [inline gdal/frmts/nitf/nitffile.c]: wrong indentation. Here and in several places below

File: gdal/frmts/nitf/nitffile.c [hunk 2]
```diff
@@ -2216,7 +2217,77 @@ static char** NITFGenericMetadataReadTREInternal(char **papszMD,
...
+				{
+                    int len = strlen(pszVal);
```
Comment (rouault) [inline gdal/frmts/nitf/nitffile.c]: all the below is over complicated. Basically everywhere in GDAL assumes that native float C type is IEEE754. So you could simplify this as approximatively:
```
float f;
memcpy(&f, pszVal, 4);
CPL_MSBPTR32(&f);
```

File: gdal/frmts/nitf/nitffile.c [hunk 3]
```diff
@@ -2216,7 +2217,77 @@ static char** NITFGenericMetadataReadTREInternal(char **papszMD,
...
+                        ((unsigned int)ncVal[3]);
+                    
+                    int s,e;
+                    long f;
+                    s = (nVal & 0x80000000UL) >> 31;
+                    e = (nVal & 0x7F800000UL) >> 23;
+                    f = (nVal & 0x007FFFFFUL);
+                    if (e == 255 && f != 0) {
+                        /* NaN (Not a Number) */
+                        sprintf(pszVal, "%f",0.0/0.0);
```
Comment (rouault) [inline gdal/frmts/nitf/nitffile.c]: overwrite pszVal is wrong here as it is a read-only variable. Use a temporary buffer

File: gdal/frmts/nitf/nitffile.c [hunk 4]
```diff
@@ -2216,7 +2217,77 @@ static char** NITFGenericMetadataReadTREInternal(char **papszMD,
...
+                    else if (e == 0 && f == 0 && s == 1) {
+                        /* Negative zero */
+                        sprintf(pszVal, "%f", 0.0);
+                    }
+                    else if (e == 0 && f == 0 && s == 0) {
+                        /* Positive zero */
+                        sprintf(pszVal, "%f",0.0);
+                    }
+				} else if (strcmp(pszType, "UINT") == 0) {
+                    int len = strlen(pszVal);
```
Comment (rouault) [inline gdal/frmts/nitf/nitffile.c]: same here. Can be simplified with something like
```
unsigned int nVal;
memcpy(&nVal, pszVal, 4);
CPL_MSBPTR32(&nVal);
```

File: gdal/frmts/nitf/nitffile.c (hunk 2 above)
Comment (crossbearer) [reply gdal/frmts/nitf/nitffile.c]: Made this change

File: gdal/frmts/nitf/nitffile.c (hunk 4 above)
Comment (crossbearer) [reply gdal/frmts/nitf/nitffile.c]: Made this change

File: gdal/frmts/nitf/nitffile.c (hunk 3 above)
Comment (crossbearer) [reply gdal/frmts/nitf/nitffile.c]: Not sure how to make this change in order to make the correct values appear.

**Comment by rouault** [comment]
Superseded by https://github.com/OSGeo/gdal/pull/2414

**Pull request closed without merging**